docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+
+Nullability information is exposed publicly as an enum, `Nullability`.
+There are 5 possible values:
+
+* `NotApplicable` - This is used when nullability information is requested about something that does not have a nullability, such as a statement, directive, or other similar constructs.
+* `NotAnalyzed` - This is used to fulfill priority #2.
+If the user requests information while asking for nullability to not be calculated, we'll use this to stand in for nullability.
+* `Unknown` - This is used for legacy C# code that does not know about nullability.
+This can appear in modern C# code that calls unannotated C#, or in certain generic variance scenarios where a common nullability cannot be inferred from the supplied arguments.
+For use in this document unknown is represented with the `~` character, but this character does not appear in source.
+* `NotNull` - This is used when we know the nullability is not null.
+* `MayBeNull` - This is used when we know the nullability may be null.
+
+Nullability information will be retrieved from the `SemanticModel`, much like type information.
+`GetSemanticModel` will return a `SemanticModel` that is aware of nullability information, and will force binding and calculation of nullability information when queried.
+We will offer a method off of `SemanticModel`, `GetSingleStatementSemanticModel()`, which will return a `SemanticModel` that does not calculate nullability information, for particularly performance-sensitive scenarios.
+This `SemanticModel` will share a cache with its parent, much like speculative `SemanticModel`s, and all nullabilities accessed on symbols retrieved from this `SemanticModel` will be `NotAnalyzed`.
+
+Generally speaking, when you want to determine the nullability of an `ITypeSymbol`, you need to check the containing context.
+The declaring symbol (e.g. `ILocalSymbol`, `IFieldSymbol`, etc.) will have the declared nullability of the type.
+`GetTypeInfo` will give you both the `ITypeSymbol` and the `Nullability` of an expression.
+Nested nullabilities for type parameters are contained on the containing `ITypeSymbol`.
+For example, if you wanted to know whether an array's contents are nullable, you will need to check the `IArrayTypeSymbol` for `ElementNullability`.
+
+Value types are always considered `NotNull`.
+Nullable value types and reference types have their state tracked, and are either `NotNull` or `MayBeNull` depending on flow state.
+Unconstrained generic parameters are being tracked, but what they will return for flow state information before explicit null checks occur in source code has not yet been decided.
+
+### Declared Nullability
+
+Declared nullability is the nullability that was explicitly typed in the source code by the programmer.
+There are some cases where this is inferred, such as lambda arguments/return types, `var` variables, and type parameter substitution.
+For these cases, flow state is necessary to compute the declared nullability.
+For example:
```

Comment:
i find this a bit hard to parse.  "explicitly typed in code" "some cases where this is inferrred".  If it's inferred, it doesn't seem like it was explicitly typed in code.

Reply:
I'll add a `usually` qualifier. Declared nullability _is_ usually explicitly typed in code, but we can infer it wherever we infer a type.